design/ozone.md

```diff
@@ -0,0 +1,246 @@
+Ozone uses `StatefulSet` for the master servers (Storage Container Manager, Ozone Manager, S3 gateway, Recon Server) and `DaemonSet` for datanodes
+
+![Ozone components](https://i.imgur.com/e7Tmij4.png)
+
+In a real world setup Ozone use the following components:
+
+ * Prometheus to collect all of the metrics
+ * Grafana to display the default dashboards
+
+## Storage requirements
+
+Similar to other Software Defined Storage, Ozone is the gateway between the local storage and network storage. 
+
+It's suggested to use local storage. Ozone suggests to use SSD or NVMe, for the metadata store and spinning disk for the real data.
+
+## Implementation
+
+The initial implementation of the Operator should be based on the available examples. The main beahviour will be similar to the minio operator (alpha), it will install Ozone cluster based on a custom CRD. Some notable differences.
+
+ * Versioning support (see the next section)
+ * Kubernetes resources are generated from the original k8s template files provided by the Ozone project (it is an advanced version of the existing ceph/csi example)
+
+### Version support
+
+By default, Ozone tools are not required by the operator. The only reason is to use the `apache/ozone` container image as the base image is to have all the required k8s resources templates.
+
+The version of Ozone can be decoupled from the version of the Operator similar to the [CSI version decoupling](https://github.com/rook/rook/blob/master/design/decouple-ceph-version.md). The default version will be defined by the rook operator version but other version can be used based on CRD entry. 
+
+### Kubernetes resource generation
+
+Ozone provides examples K8s resources files and transformation definitions using [flekszible](https://github.com/elek/flekszible). This is a composition based k8s resource generator where the transformations can be reused.
+
+Let's say we have a simple kubernetes resource:
+
+```yaml
+apiVersion: apps/v1
+kind: DaemonSet
+metadata:
+  name: datanode
+  labels:
+    app.kubernetes.io/component: ozone
+spec:
+  selector:
+    matchLabels:
+      app: ozone
+      component: datanode
+  template:
+    metadata:
+      labels:
+        app: ozone
+        component: datanode
+    spec:
+      containers:
+      - name: datanode
+        image: "apache/hadoop:0.5.0-SNAPSHOT"
+        args: ["ozone","datanode"]
+        ports:
+           - containerPort: 9870
+             name: rpc
+```
+
+Prometheus annotations can be added by a simple rule:
+
+```yaml
+type: Add
+path: ["spec","template","spec","metadata","annotations"]
+trigger:
+   metadata:
+      name: datanode
+value:
+   prometheus.io/scrape: "true"
+   prometheus.io/port: "9882"
+   prometheus.io/path: "/prom"
+```
+
+This transformation can be applied to any of the k8s resources.
+
+Until now it is very similar to Kustomize, but here we have more predefined transformations (DaemonSet to StatefulSet converter, NodePort service generator, value replacer).
```

Comment:
There is a converter from DaemonSet to StatefulSet? That is a fundamental change in resource type. Can you provide more details about that scenario?

Reply:
Until now we used DaemonSet by default for datanodes as it was more realistic (usually we need one store node per physical nodes). But for Minikube the daemonset is transformed to statefulset (yes, the semantic has a fundamental difference but the syntax is very similar. And it worked well).

As I wrote earlier we decided to use StatefulSet for datanodes (end emulates the daemonset property with anti-affinity rules). Mainly to make it easier to use PVCs. So this conversion is not a good example any more.